lib/g_searcher/reports.ex

```diff
@@ -0,0 +1,88 @@
+defmodule GSearcher.Reports do
+  import Ecto.Query, warn: false
+
+  alias Ecto.Multi
+  alias GSearcher.Repo
+  alias GSearcher.Search
+  alias GSearcher.Search.Report
+
+  NimbleCSV.define(CSVParser, separator: "\t", escape: "\"")
+
+  def create_report(user_id, title, file_path) do
+    Multi.new()
+    |> Multi.insert(
+      :report,
+      Report.changeset(%{title: title, csv_path: file_path, user_id: user_id})
+    )
+    |> Multi.run(:search_keywords, fn _, %{report: %{id: report_id, csv_path: csv_path}} ->
+      save_keywords_from_file(csv_path, report_id)
+    end)
+    |> Repo.transaction()
+    |> handle_report_transaction_response()
+  end
+
+  defp save_keywords_from_file(csv_path, report_id) do
+    with {:ok, csv_path} <- validate_file_path(csv_path),
+         {:ok, file_stream} <- stream_csv(csv_path),
+         {:ok, keyword_list} <- save_keywords(file_stream, report_id) do
+      case Enum.member?(keyword_list, :error) do
```

Comment:
If one of the keywords failed to save. Do we need to rollback the other created keywords? Since this function returns `:error` when one of the keywords error.

Reply:
yeah, that's the plan, I was thinking rolling back would just be better and display on the frontend that the CSV failed to be properly parsed than having an incomplete report. 🤔 

My idea was that if a user uploads a CSV but halfway through the file it gets malformed or incorrect formatting. Since we're streaming the file, I'm thinking it would probably be easier to fail the entire file than handling the successful entries and displaying each failure 🤔 

What do you think about it? @hanam1ni

Reply:
I agree with your approach to fail the entire file and the rollback part maybe you can use the `transaction/2` function https://hexdocs.pm/ecto/Ecto.Repo.html#c:transaction/2 🙌

Reply:
Yeah, great idea 👍 , I'll use that here.